### PR TITLE
Fixes #12517: Add generic methods to assign monitoring templates and parameters to nodes

### PR DESCRIPTION
**Pull request by VictorQrt**
https://www.rudder-project.org/redmine/issues/12517

### CODE REVIEW

File: tree/30_generic_methods/monitoring_parameter.cf [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Monitoring parameter
+# @description Add a monitoring parameter to a node
```
Comment (amousset) [inline tree/30_generic_methods/monitoring_parameter.cf]: We could use a `@documentation` section here, with a usage example. It should explain what a monitoring parameter is and it is used for. See other methods for syntax (the content itself is in matkdown)

File: tree/30_generic_methods/monitoring_template.cf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Monitoring template
+# @description Add a monitoring template to a node
```
Comment (amousset) [inline tree/30_generic_methods/monitoring_template.cf]: same here

File: tree/30_generic_methods/monitoring_parameter.cf [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#####################################################################################
+# Copyright 2013 Normation SAS
```
Comment (amousset) [inline tree/30_generic_methods/monitoring_parameter.cf]: 2018

File: tree/30_generic_methods/monitoring_template.cf [hunk 4]
```diff
@@ -0,0 +1,48 @@
+#####################################################################################
+# Copyright 2013 Normation SAS
```
Comment (amousset) [inline tree/30_generic_methods/monitoring_template.cf]: same here

**Comment by VictorQrt** [comment]
Commit modified

File: tree/30_generic_methods/monitoring_parameter.cf (hunk 1 above)
Comment (peckpeck) [reply tree/30_generic_methods/monitoring_parameter.cf]: add a (requires a monitoring plugin)

File: tree/30_generic_methods/monitoring_template.cf (hunk 2 above)
Comment (peckpeck) [reply tree/30_generic_methods/monitoring_template.cf]: add a (requires a monitoring plugin)

File: tree/30_generic_methods/monitoring_parameter.cf [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+# @name Monitoring parameter
+# @description Add a monitoring parameter to a node
+#
+# @documentation
+# This method adds monitoring parameters to rudder nodes. The monitoring parameters are used to
+# pass configuration to the monitoring plugins running with Rudder. Expected keys and parameters
+# are specific to each plugin and can be found in their respective documentation.
+# 
+#
+# @parameter key Name of the key
```
Comment (peckpeck) [inline tree/30_generic_methods/monitoring_parameter.cf]: Name of the parameter

File: tree/30_generic_methods/monitoring_parameter.cf [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    "args" slist => {"${key}", "${value}"}; 
+
+  classes:
+  
+    "should_report" expression => "${report_data.should_report}";
+
+  methods:
+    
+    "disable_reporting_${old_class_prefix}" usebundle => disable_reporting;
+    "Write csv file" usebundle => file_ensure_lines_present("/var/rudder/tmp/rudder_monitoring.csv", "param,${key},${value}");
```
Comment (peckpeck) [inline tree/30_generic_methods/monitoring_parameter.cf]: Is the case where value contains a comma handled properly ?

**Comment by VictorQrt** [comment]
Commit modified

**Comment by VictorQrt** [comment]
Commit modified

File: tree/30_generic_methods/monitoring_parameter.cf (hunk 6 above)
Comment (VictorQrt) [reply tree/30_generic_methods/monitoring_parameter.cf]: Yes, the plugins will escape the commas past the second and the parameter will contain it as expected.

File: tree/30_generic_methods/monitoring_parameter.cf [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+# @description Add a monitoring parameter to a node (requires a monitoring plugin)
+#
+# @documentation
+# This method adds monitoring parameters to rudder nodes. The monitoring parameters are used to
+# pass configuration to the monitoring plugins running with Rudder. Expected keys and parameters
+# are specific to each plugin and can be found in their respective documentation.
+# 
+#
+# @parameter key Name of the parameter
+# @parameter value Value of the key
```
Comment (peckpeck) [inline tree/30_generic_methods/monitoring_parameter.cf]: value of the parameter

**Comment by VictorQrt** [comment]
Commit modified

**Pull request merged**